libs/fujitsu-thermal-printer/src/printing.ts

```diff
@@ -0,0 +1,351 @@
+
+  return new Int8Array(compressedData);
+}
+
+export function compressBitImage(
+  uncompressed: UncompressedBitImage
+): CompressedBitImage {
+  debug('compressing bit image');
+  return {
+    height: uncompressed.height,
+    data: packBitsCompression(uncompressed.data),
+    compressed: true,
+  };
+}
+
+const WAIT_FOR_BUFFER_NOT_FULL_TIMEOUT_MS = 2.5 * 1000;
+const WAIT_FOR_BUFFER_FLUSH_TIMEOUT_MS = 10 * 1000;
+
+export async function printPageBitImage(
+  driver: FujitsuThermalPrinterDriverInterface,
+  compressedBitImages: IteratorPlus<CompressedBitImage>
+): Promise<Result<void, RawPrinterStatus>> {
+  // print all bit images that compose the page
+  await driver.setReplyParameter(PRINT_ONGOING_REPLY_PARAMETER);
+  for (const compressedBitImage of compressedBitImages) {
+    // wait for the buffer to be ready before sending another bit image. this
+    // is not strictly necessary - either the OS or `node-usb` layer handles
+    // buffering - but it allows us to keep our execution roughly in sync with
+    // the hardware and allows us to bail earlier if the printer stops
+    const waitForPrintReadyResult = await waitForPrintReadyStatus(driver, {
+      interval: 100,
+      timeout: WAIT_FOR_BUFFER_NOT_FULL_TIMEOUT_MS,
+      replyParameter: PRINT_ONGOING_REPLY_PARAMETER,
+    });
+    if (waitForPrintReadyResult.isErr()) {
+      return waitForPrintReadyResult;
+    }
+
+    assert(compressedBitImage);
+    driver.printBitImage(compressedBitImage);
+  }
+
+  await driver.setReplyParameter(PRINT_PROCESSING_REPLY_PARAMETER);
+  const waitForPrintFinished = await waitForPrintReadyStatus(driver, {
+    interval: 100,
+    timeout: WAIT_FOR_BUFFER_FLUSH_TIMEOUT_MS,
+    replyParameter: PRINT_PROCESSING_REPLY_PARAMETER,
+  });
+  if (waitForPrintFinished.isErr()) {
+    return waitForPrintFinished;
+  }
+
+  debug('printed page successfully');
+  return ok();
+}
+
+export async function print(
+  driver: FujitsuThermalPrinterDriverInterface,
+  pdfData: Uint8Array
+): Promise<Result<void, RawPrinterStatus>> {
+  const pdfImages = pdfToImages(Buffer.from(pdfData), {
+    scale: 200 / 72,
+  });
+  for await (const { page, pageNumber, pageCount } of pdfImages) {
+    debug(`printing page ${pageNumber} of ${pageCount}...`);
+    debug(`page dimensions: ${page.width} x ${page.height}`);
+    const printPageResult = await printPageBitImage(
+      driver,
+      iter(trimAndChunkImageData(page))
+        .map((imageData) => imageDataToBinaryBitmap(imageData, {}))
+        .map(bitmapToBitImage)
+        .map(compressBitImage)
+    );
+    if (printPageResult.isErr()) {
+      await driver.setReplyParameter(IDLE_REPLY_PARAMETER);
+      return printPageResult;
+    }
+  }
+
+  await driver.setReplyParameter(IDLE_REPLY_PARAMETER);
+  return ok();
+}
+
+export async function printFixture(
+  pdfFixturePath: string,
+  driver: FujitsuThermalPrinterDriver
+): Promise<void> {
+  const pdfData: Uint8Array = readFileSync(pdfFixturePath);
```

Comment:
nit: Should we be moving to the `readFileSync` util from `libs/fs` that Brian added? This looks like it's not in the main app path, so maybe less important